### PR TITLE
refactor(storage): adapt time travel cache to partial checkpoint

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

After partial checkpoint, the time travel's hummock version cache should use `(table_id, epoch)` as key rather than original `epoch`.
**The cache is necessary because a batch query may trigger many iterators in a compute node due to multiple vnodes.**
This PR also optimizes the lock used in cache implementation.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/time_travel_version_cache.rs [hunk 1]
```diff
@@ -13,43 +13,65 @@
...
+use futures::FutureExt;
 use moka::sync::Cache;
+use parking_lot::{Mutex, RwLock};
 use risingwave_hummock_sdk::HummockEpoch;
-use tokio::sync::Mutex;
 
 use crate::hummock::local_version::pinned_version::PinnedVersion;
 use crate::hummock::HummockResult;
 
+type InflightRequest = Shared<Pin<Box<dyn Future<Output = HummockResult<PinnedVersion>> + Send>>>;
```
Comment (wenym1) [inline src/storage/src/hummock/time_travel_version_cache.rs:28]: Great to know this useful `Shared` future.

With this `Shared` future, I think we don't need to separate the ready `PinnedVersion` and the `InflightRequest`. They can be unified to be having a single `moka::sync::Cache<(TableId, HummockEpoch), InflightResult>>`.

The logic can be as simple as 
```
pub async fn get_or_insert(
    &self,
    table_id: u32,
    epoch: HummockEpoch,
    fetch: impl Future<Output = HummockResult<PinnedVersion>> + Send + 'static,
) -> HummockResult<PinnedVersion> {
    self.cache.entry((epoch, table_id)).or_insert_with(|| fetch.boxed().shared()).value().clone().await
}
```

Comment (zwang28) [reply src/storage/src/hummock/time_travel_version_cache.rs]: Thanks.

File: src/storage/src/hummock/time_travel_version_cache.rs [hunk 2]
```diff
@@ -62,11 +40,25 @@ impl SimpleTimeTravelVersionCacheInner {
...
+        fetch: impl Future<Output = HummockResult<PinnedVersion>> + Send + 'static,
+    ) -> HummockResult<PinnedVersion> {
+        let result = self
+            .cache
+            .entry((table_id, epoch))
+            .or_insert_with(|| fetch.boxed().shared())
+            .value()
+            .clone()
+            .await;
+        if result.is_err() {
```
Comment (wenym1) [inline src/storage/src/hummock/time_travel_version_cache.rs:56]: In terms of error handling, we can handle the the case of error in an atomic way with [or_insert_with_if](https://docs.rs/moka/latest/moka/sync/struct.OwnedKeyEntrySelector.html#method.or_insert_with_if) of the entry. 

Basically we want to reset the cache entry with a new future when we know that, first, the original future is ready, and second, the output of the original future is error. We can use `if let Some(result) = future.peek() && result.is_err()` to detect the case.

Comment (zwang28) [reply src/storage/src/hummock/time_travel_version_cache.rs]: Thanks

**Review by wenym1 — APPROVED** [review]
LGTM!

**Pull request merged**